### PR TITLE
os/bluestore: proper locking for BlueFS prefetching

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -1477,6 +1477,10 @@ int BlueFS::_read(
     if (off < buf->bl_off || off >= buf->get_buf_end()) {
       s_lock.unlock();
       std::unique_lock u_lock(h->lock);
+      if (off >= buf->bl_off && off < buf->get_buf_end()) {
+        // if precondition has changed during locking upgrade.
```
Comment (liewegas) [inline src/os/bluestore/BlueFS.cc]: need s_lock.lock() before the continue here, right?

Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: no, we're under exclusive locking for the same h->lock mutex here. More than enough IMO..

Comment (liewegas) [reply src/os/bluestore/BlueFS.cc]: the continue will mean that u_lock goes out of scope and we drop the lock, though...

Comment (ifed01) [reply src/os/bluestore/BlueFS.cc]: yeah, my bad.. fixed.

**Pull request merged**